server/apps/publish/formatters/aap_ipnews_formatter.py

```diff
@@ -30,26 +31,47 @@ def format(self, article, output_channel, selector_codes):
             pub_seq_num = superdesk.get_resource_service('output_channels').generate_sequence_number(output_channel)
 
             odbc_item = {}
-            odbc_item['originator'] = article.get('originator', None)
+            odbc_item['originator'] = article.get('source', None)
             odbc_item['sequence'] = pub_seq_num
             odbc_item['category'] = article.get('anpa-category', {}).get('qcode')  # @category
-            odbc_item['headline'] = article.get('headline', '')  # @headline
-            odbc_item['author'] = article.get('byline', '')  # @author
-            odbc_item['keyword'] = article.get('slugline', None)
+            odbc_item['headline'] = article.get('headline', '').replace('\'', '\'\'')  # @headline
+            odbc_item['author'] = article.get('byline', '').replace('\'', '\'\'')  # @author
+            odbc_item['keyword'] = article.get('slugline', None).replace('\'', '\'\'')
 
-            if article['subject'][0]:
-                odbc_item['subject_reference'] = article['subject'][0].get('qcode', None)
+            if article['subject'][0] and 'qcode' in article['subject'][0]:
+                odbc_item['subject_reference'] = article['subject'][0].get('qcode', '        ')
                 if odbc_item['subject_reference']:
                     odbc_item['subject'] = subject_codes[odbc_item['subject_reference'][:2] + '000000']
-                    odbc_item['subject_matter'] = subject_codes[odbc_item['subject_reference'][:5] + '000']
-                    odbc_item['subject_detail'] = subject_codes[odbc_item['subject_reference']]
+                    if odbc_item['subject_reference'][2:5] != '000':
+                        odbc_item['subject_matter'] = subject_codes[odbc_item['subject_reference'][:5] + '000']
+                    else:
+                        odbc_item['subject_matter'] = ''
+                    if not odbc_item['subject_reference'].endswith('000'):
+                        odbc_item['subject_detail'] = subject_codes[odbc_item['subject_reference']]
+                    else:
+                        odbc_item['subject_detail'] = ''
+            else:
+                odbc_item['subject_reference'] = '        '
+                odbc_item['subject'] = ''
+                odbc_item['subject_matter'] = ''
+                odbc_item['subject_detail'] = ''
 
             odbc_item['take_key'] = article.get('anpa_take_key', None)  # @take_key
             odbc_item['usn'] = article.get('unique_id', None)  # @usn
             if article['type'] == 'preformatted':
                 odbc_item['article_text'] = article.get('body_html', '').replace('\'', '\'\'')  # @article_text
             elif article['type'] == 'text':
-                text = BeautifulSoup(article.get('body_html', '')).text
+                soup = BeautifulSoup(article.get('body_html', ''))
+                text = ''
```

Comment:
Better to use a string builder (i.e. `io.StringIO`) or an accumulator list to build a string from many smaller fragments.